homeassistant/helpers/icon.py

```diff
@@ -1,7 +1,188 @@
+
+    async def async_fetch(
+        self,
+        category: str,
+        components: set[str],
+    ) -> list[dict[str, dict[str, Any]]]:
+        """Load resources into the cache."""
+        if components_to_load := components - self.loaded:
+            await self._async_load(components_to_load)
+
+        return [
+            self.cache.get(component, {}).get(category, {}) for component in components
+        ]
+
+    async def _async_load(self, components: set[str]) -> None:
+        """Populate the cache for a given set of components."""
+        _LOGGER.debug(
+            "Cache miss for: %s",
+            ", ".join(components),
+        )
+
+        integrations: dict[str, Integration] = {}
+        domains = list({loaded.rpartition(".")[-1] for loaded in components})
+        ints_or_excs = await async_get_integrations(self.hass, domains)
+        for domain, int_or_exc in ints_or_excs.items():
+            if isinstance(int_or_exc, Exception):
+                raise int_or_exc
+            integrations[domain] = int_or_exc
+
+        icons = await _async_get_component_icons(self.hass, components, integrations)
+
+        self._build_category_cache(components, icons)
+
+        self.loaded.update(components)
+
+    @callback
+    def _build_category_cache(
+        self,
+        components: set[str],
+        icons: dict[str, dict[str, Any]],
+    ) -> None:
+        """Extract resources into the cache."""
+        resource: dict[str, Any] | str
+        categories: set[str] = set()
+        for resource in icons.values():
+            categories.update(resource)
+
+        for category in categories:
+            new_resources = build_resources(icons, components, category)
+            for component, resource in new_resources.items():
+                category_cache: dict[str, Any] = self.cache.setdefault(
+                    component, {}
+                ).setdefault(category, {})
+
+                if isinstance(resource, dict):
+                    category_cache |= recursive_flatten(
+                        f"component.{component}.{category}.",
+                        resource,
+                    )
+                else:
+                    category_cache[f"component.{component}.{category}"] = resource
+
+
+async def async_get_icons(
+    hass: HomeAssistant,
+    category: str,
+    integrations: Iterable[str] | None = None,
+) -> dict[str, Any]:
+    """Return all icons of integrations.
+
+    If integration specified, load it for that one; otherwise default to loaded
+    intgrations.
+    """
+    lock = hass.data.setdefault(ICON_LOAD_LOCK, asyncio.Lock())
+
+    if integrations is not None:
+        components = set(integrations)
+    else:
+        components = set(hass.config.components)
+
+    async with lock:
+        if ICON_FLATTEN_CACHE in hass.data:
+            cache = hass.data[ICON_FLATTEN_CACHE]
+        else:
+            cache = hass.data[ICON_FLATTEN_CACHE] = _IconsCache(hass)
```

Comment:
```suggestion
        cache = hass.data.setdefault(ICON_FLATTEN_CACHE, _IconsCache(hass))
```

Reply:
This is no the same, this will create a new cache instance to pass into the method everytime this part of the code is called (while it is being omitted constantly, as it already existed). This is a side-effect from passing it in an an argument.